### ISSUE TITLE
Gmail clips emails at 102 kB or with special characters

### ISSUE THREAD

**Issue by hteumeuleu** [hteumeuleu/email-bugs#41]
Yesterday, an interesting conversation was started on the [#emailgeeks Slack](https://email.geeks.chat/) by @cossssmin regarding Gmail clipping limit.

> Do we have a source for the magic 102 kB message size Gmail clipping limit? I'm curious if someone actually tested this or if it was communicated from an official source, as I've been blindly following it (like most of us probably have), and I've recently seen emails that were 99.something kB getting clipped (before you ask, it was _after_ ESP added stuff to it).

A few people, including myself, weighed in to share their experience and a few test results, but without a definitive answer yet. I loved the little collaborative investigation that went on for a few hours, but it seems Slack is not really appropriate for this (given the temporary nature of conversations there). So I thought here would be a good place to continue this together.

# The problem

!["Message clipped" screenshot in Gmail](https://i.imgur.com/P8pOCix.jpg)

When HTML emails are too large, Gmail clips them with a `[Message clipped]` notice and a `View entire message` link. It's been widely shared that this clipping occurs after 102 kB (example: [Gmail is clippin my email](https://kb.mailchimp.com/delivery/deliverability-research/gmail-is-clipping-my-email) on Mailchimp). But no one seems to know where this number comes from. And different people have experience different results around the 100 kB mark.

So what is going on exactly? Can we figure out Gmail's clipping algorithm?

**Comment by revelt** [comment]
Thank you for sharing!

It's definitely not exactly `100KB, I've seen crop happen at lesser sizes. My "rule of thumb" has always been to aim at file sizes less than `80KB`. 1 character = 1 byte so that's around `80,000` characters, what's easy to check in the code editor if you select-all and see the status bar for total character count.

What's also interesting, if you consider, Gmail will **receive** not your HTML but what ESP sent it, basically, what you see in the "message's raw source". Various factors will bloat the served HTML code in there: ESP link URL scrambling, quoted printable encoding, sometimes ESP's serve message as BASE64-encoded...

So, maybe `100` could be _the_ threshold, but definitely uppermost and very possible in the shape of 100x1024 characters in the raw source in the HTML part of the message, _as received by email server_. But I'd aim for less than `80,000` characters.

**Comment by hellocosmin** [comment]
Here's the third test file in Windows:

![image](https://user-images.githubusercontent.com/1656595/40177163-6cefff92-59e6-11e8-9b2a-8a9e880f1c11.png)

Differences are to be expected, and I think neither are true to what Gmail actually counts server side, as @revelt pointed out.

On this note, I'd encourage taking tools that test your HTML file size against 'Gmail's limit' with a grain of salt. The point is 'we don't know _exactly_ yet', so [apparently useful tools](https://www.sendwithus.com/resources/gmail_size_test) might be a little misleading. It can happen they're just a tiny bit off, but that can be the difference between your tracking pixel being removed in all Gmail clients, or not. 

Here's the thrid test file that was clipped, in the tool I linked to above:

![image](https://user-images.githubusercontent.com/1656595/40177708-006e4188-59e8-11e8-8525-fe744dbc7a59.png)

**Comment by revelt** [comment]
Very very cheeky

**Comment by hteumeuleu** [comment]
@hellocosmin You tested the file after clipping. So isn't the tool actually accurate to show that it will pass Gmail clipping? If I test the original file, it indeed says the email is too big (at _ 175.869140625 Kb
_).

@revelt Good point on being careful with manipulations done on the ESP side. I used [Putsmail](http://putsmail.com/) for all my tests, which is pretty safe as far as I know.

**Comment by hellocosmin** [comment]
Indeed Rémi, I've tested the same third _result_ file. You mentioned macOS reported to be 102KB on disk. 

So an HTML that you'd think would very likely be clipped (according to macOS' report) was actually reported as 'safe'. Even if we don't consider 'on disk', and we take it to be 100KB, the difference is still large enough to be misleading: there's a 2,1318359375KB difference at the very least, so logically I can imagine a ~103.13KB file as also being reported 'safe'.

My point was we should take such file size weighing tools as estimates, and definitely not fully trust some JavaScript that calculates file size as being the same thing Gmail does when deciding to clip :)

**Comment by pbiolsi** [comment]
From your testing/understanding, what do we know about how/if images impact the calculable email weight by Gmail? 

Since we're all serving images through some external server or CDN, I've always assumed that their size only had an implication on bandwidth/loadtime... but perhaps that is incorrect and the Gmail image cache is at all to blame for clipping? 

Our embedded styles run pretty lean (so not triggering the CSS characters limit) and we use built-in minification provided by our ESP (Listrak) at the time of send to remove whitespace. Yet, we still see emails as small as 32kb getting clipped by Gmail.

Any possibility they impose a pixel height-based limitation that could be triggering that? Like Outlook has been known to?

**Comment by hellocosmin** [comment]
Never saw emails that small being clipped in Gmail, and I send quite a few that were larger than 32KB. You sure your ESP isn't adding stuff to what Gmail receives?

**Comment by revelt** [comment]
@pbiolsi check the raw source. I'm pretty sure Gmail is measuring raw source. Now, it's message is multipart, there's Base64 with some heavy link scrambling, sizes will bloat significantly..

1 character in your email's raw source = 1 byte. That includes escape characters (used in quoted-printable encoding for example). After you rule-out the raw source, then move to images.

Sizes as low as 32KB should not get clipped, something's wrong here

**Comment by pbiolsi** [comment]
@hellocosmin @revelt great points about the ESP bloat, however I've actually seen this in Litmus previews using the Chrome extension (so serving local source pre-ESP... also un-minified in this case). Maybe something else is causing this from the Litmus side, but I believe those previews use PutsMail so should be pretty true to source.

I'll investigate more and post back.

**Comment by hteumeuleu** [comment]
Last week on the [#emailgeeks Slack](https://email.geeks.chat/), @M-J-Robbins shared an interesting example that gets clipped because of a special character.

> Just got an email through that Gmail said was clipped but the code is tiny and no obvious errors.
>
> This is the character in question `` (not sure if slack will auto convert that) but I believe it’s this one https://symbl.cc/en/0092/>
> Here’s an html example https://litmus.com/scope/ajulzityfxh6

```html
<html>
  <p>youre</p>
```
![A screenshot on Gmail showing an email getting clip because of a special character](https://user-images.githubusercontent.com/3451753/47026074-edb81880-d164-11e8-87ea-1c68468a2105.png)

This only happens on the new Gmail redesign, not on the old one. This is very interesting because I also noticed a lot of emails getting clipped for no apparent reason since the redesign. I'll try to see if there are more characters triggering this.

**Comment by ericlepetit** [comment]
Funny enough the Github email notification for this message was clipped on Gmail :)
I will bring this up to the Gmail team.

**Comment by revelt** [comment]
hi all, just crossposting here from email geeks slack for posterity because this is surfacing up once in a while. If you want to check, does your email contain any non-ascii characters (like Unicode's culprit "Private Use Two" above), I created a CLI app (terminal app) for that: https://www.npmjs.com/package/email-all-chars-within-ascii-cli

**Comment by hthompson82** [comment]
Not only have I had issues with emails smaller than 102kb clipping, I've also sent several emails WELL over the 102kb (180-220kb) which _don't_ clip. Anyone have any ideas? Getting a brick wall from Google whenever I try to talk to anyone about it.

**Comment by revelt** [comment]
@hthompson82 hi! Do you remember, what was the encoding of your raw messages that arrived into Gmail server (for example, quoted printable, Base64 etc.); did you measure character count in the raw HTML, decoded or the HTML that was put into ESP (former is more interesting); also was the message multipart and if so, how big was text part (hypothesis being text version's content might have affected Gmail clipping limit)?

**Comment by hthompson82** [comment]
@revelt The encoding is "text/html; charset=utf-8". I count the kb size according to how it comes in to my Outlook (since I can't see size in Gmail) so that I can be sure it includes encoding, dynamic data etc. Our test sends are generally multipart but the text version varies between one word ("test") and a couple of paragraphs. Testing without the multipart text version still results in clipping, on the ones which clip.

**Comment by jkupczak** [comment]
@hthompson82 When you say Outlook, do you mean the Outlook web client or do you mean the desktop application that you have to install? 

If you mean the desktop application, does Outlook let you see the original message? I don't have Outlook installed at the moment so I can't check for myself. But I thought that Outlook would only show you the message source which is the version that Outlook parsed using the Word engine.

**Comment by hellocosmin** [comment]
@jkupczak the desktop Outlook app can show you the original HTML source, as it was received. You need to double click the message to open in a new window, then click "Message" in the top left, and then:

![View source in Outlook](https://user-images.githubusercontent.com/1656595/50305772-8d16c600-049c-11e9-8e37-da22c17061db.png)

**Comment by hthompson82** [comment]
> @hthompson82 When you say Outlook, do you mean the Outlook web client or do you mean the desktop application that you have to install?
> 
> If you mean the desktop application, does Outlook let you see the original message? I don't have Outlook installed at the moment so I can't check for myself. But I thought that Outlook would only show you the message source which is the version that Outlook parsed using the Word engine.

Hi @jkupczak , i use the Outlook desktop app, which includes the size as a column in the inbox. (I also test emails via mail.yahoo.com and hotmail.com in the browser)

As an aside, I've also tested compressing the HTML (stripping out all white space) but to no avail.

**Comment by hteumeuleu** [comment]
> Last week on the [#emailgeeks Slack](https://email.geeks.chat/), @M-J-Robbins shared an interesting example that gets clipped because of a special character.
> 
> > Just got an email through that Gmail said was clipped but the code is tiny and no obvious errors.
> > This is the character in question `` (not sure if slack will auto convert that) but I believe it’s this one https://unicode-table.com/en/0092/
> > Here’s an html example https://litmus.com/scope/ajulzityfxh6
> 
> ```
> <html>
>   <p>�you�re</p>
> ```
> 
> ![A screenshot on Gmail showing an email getting clip because of a special character](https://user-images.githubusercontent.com/3451753/47026074-edb81880-d164-11e8-87ea-1c68468a2105.png)
> 
> This only happens on the new Gmail redesign, not on the old one. This is very interesting because I also noticed a lot of emails getting clipped for no apparent reason since the redesign. I'll try to see if there are more characters triggering this.

This bug mentioned in this thread seems fixed. Can anyone else confirm?

**Comment by revelt** [comment]
Not fixed. Same thing — both original `<html><p>youre</p>` and also same thing wrapped with normal HTML head/body in a table. This is still happening.

![screen shot 2019-02-12 at 00 40 26](https://user-images.githubusercontent.com/8344688/52603501-03adbf80-2e5f-11e9-8f42-4a0b1abdb36b.png)

**Comment by hteumeuleu** [comment]
Got an even shorter example triggering the clipped message being shown because of a special character:

```
<html>
  <p>©</p>
```
Encoding the copyright character into `&copy;` fixes the problem here.

**Comment by jclusso** [comment]
@hteumeuleu I have this issue and I'm using &copy; No idea why though.

**Comment by jkupczak** [comment]
@jclusso

Got any information you can share.

**Comment by revelt** [comment]
yeah, let's create a minimal case to be able to reproduce.. as they say "He who asserts must prove"

**Comment by jclusso** [comment]
I think the issue is that the provider I'm using to send the emails is setting the `Content-Type` header to `text/html; charset=iso-8859-1`. This still confuses me why `&copy;` won't work since it should as far as I'm aware. Most other emails I get seem to be `text/html; charset=utf-8` which makes me believe that would solve this.

**Comment by hthompson82** [comment]
Just sent two test emails from Adobe Campaign, one with &#x00A9; and one with ©. The unencoded version clipped, and the encoded one did not. 

```
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <meta name="viewport" content="width=device-width" />
<title>Untitled Document</title>
</head>

<body>
	&#x00A9;
</body>
</html>
```

**Comment by revelt** [comment]
@hthompson82 I checked the Adobe Creative Cloud newsletter which is allegedly sent from their platform. The encoding in the raw source is `charset="windows-1252"`, encoded in quoted printable, with a corresponding `charset=Windows-1252` `meta` tag in HTML. It seems all nice until one tries to decode the following raw piece:

```
See the tips<span class=3D"we=
b"> =9B</span></a></td>=20
```

It is not `Windows-1252` but `Windows-1251`/`Windows-1257` — a chevron. Test yourselves, https://dencode.com/en/string/quoted-printable

So, to sum up, it seems that Adobe Campaign platform is using a **wrong charset to encode their quoted printable**, and as a consequence, all non-ascii characters get lost. All html-encoded characters don't get encoded because they're within ascii so they're fine.

**Comment by cossssmin** [comment]
Here's a weird one:

https://m3.news.ubisoft.com/nl/jsp/m.jsp?c=%405nrrN%2BHHqfIXH1q%2FHaFADEvE%2Bv1jGn1Gp%2FMdstPVNaI%3D

19KB email, shown in full, but still showing the 'clipped' message. I checked the source in Gmail web, and even the tracking pixel is there, so nothing was actually clipped...

This is what it looks like in Gmail web:

![image](https://user-images.githubusercontent.com/1656595/82118384-c03bb280-977f-11ea-8080-df95d22cb41d.png)

**Comment by hteumeuleu** [comment]
@cossssmin There's a `©` character in the footer, so I guess that's why.

**Comment by Nivicious** [comment]
Just re-coded a template for a new client due to clipping.
Reduced the raw HTML file by 100kb to now sit at 51kb.
Tested via Putsmail and it's still clipping.

Read through this thread and simply replaced the `©` character with `&copy;` and voila, fixed.

**Comment by hteumeuleu** [comment]
I just spent half an hour wondering why I got the "Message clipped" on an email. Turns out, I had an HTML comment in french with an accented character (like `<!-- Mentions légales -->`). I removed the accent and Gmail's message disappeared. Here’s a simple test code to try: 

```
Hello world ! <!-- é -->
```

**Comment by vladh** [comment]
I encountered this problem when sending email to Gmail from Outlook. Sending a message with any German characters, such as "Grüsse", would cause the message to clip in Gmail.

I fixed this problem by making sure my email was encoded as UTF-8, which can be set in the Outlook settings.

**Comment by Achar-Bhagyashree** [comment]
Hey All, a hyphen used in the pre-header text was triggering view entire message in Gmail. While I changed this hyphen to an HTML entity issue got resolved!
Eg: "Take action – please provide your details" When I updated this to :  "Take action &ndash; please provide your details" issue is resolved.

**Comment by cseeger** [comment]
Can confirm `©` was causing this in our emails. Switching to `&copy;` fixed it.

**Comment by alicemeichi** [comment]
Just tested out both the [special character](https://symbl.cc/en/0092/) that @M-J-Robbins  referred to and the ©, and neither seem to clip anymore. Gmail must have fixed this along with the [MIME part bug](https://alicelicode.medium.com/does-gmail-clip-at-102kb-or-100kb-what-its-actually-counting-728c8c7497d2) I noticed before.
